subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/DefaultCurrentFileCollectionFingerprint.java

```diff
@@ -36,25 +39,45 @@
     private final String identifier;
     private final FileSystemSnapshot roots;
     private final ImmutableMultimap<String, HashCode> rootHashes;
+    private final HashCode strategyConfigurationHash;
     private HashCode hash;
 
     public static CurrentFileCollectionFingerprint from(FileSystemSnapshot roots, FingerprintingStrategy strategy) {
+        return from(roots, strategy, null);
+    }
+
+    public static CurrentFileCollectionFingerprint from(FileSystemSnapshot roots, FingerprintingStrategy strategy, @Nullable  FileCollectionFingerprint candidate) {
         if (roots == FileSystemSnapshot.EMPTY) {
             return strategy.getEmptyFingerprint();
         }
-        Map<String, FileSystemLocationFingerprint> fingerprints = strategy.collectFingerprints(roots);
+
+        ImmutableMultimap<String, HashCode> rootHashes = SnapshotUtil.getRootHashes(roots);
+        Map<String, FileSystemLocationFingerprint> fingerprints;
+        if (candidate != null && candidate.getStrategyConfigurationHash().equals(strategy.getConfigurationHash()) && Iterables.elementsEqual(candidate.getRootHashes().entries(), rootHashes.entries())) {
```

Comment:
This condition is too complicated. A few ideas:

- We can do away with `candidate != null`.
- I think we shouldn't have `getStrategyConfigurationHash()` exposed, but instead have a specific method to check the strategy, e.g. something like `wasCreatedWithStrategy(FingerprintingStrategy)`.
- Why not use `candidate.getRootHashes().equals(rootHashes)`?
- Consider breaking up the condition into multiple lines.

Reply:
I started out with `candidate.getRootHashes().equals(rootHashes)`, though that doesn't work since the equals on `Map` does not take the order into account. Though here the order is important, at least for classpath fingerprinters.

Reply:
If the order matters for some strategies, then it seems the strategy should be responsible for comparing the hashes, no?

Reply:
Addressed in https://github.com/gradle/gradle/pull/17789.